tokio/src/sync/task/atomic_waker.rs

```diff
@@ -178,8 +182,16 @@ impl AtomicWaker {
         {
             WAITING => {
                 unsafe {
+                    // If `into_waker` panics (because it's code outside of
+                    // AtomicWaker) we need to prime a guard that is called on
+                    // unwind restore the waker to a WAITING state. Otherwise
+                    // any future calls to register will incorrectly be stuck in
+                    // a state where it believes it's being updated by someone
+                    // else.
+                    let guard = PanicGuard(&self.state);
                     // Locked acquired, update the waker cell
                     self.waker.with_mut(|t| *t = Some(waker.into_waker()));
+                    std::mem::forget(guard);
```

Comment:
I think this is slightly easier to read.
```suggestion
                    let guard = PanicGuard(&self.state);
                    let new_waker = waker.into_waker();
                    std::mem::forget(guard);
                    // Locked acquired, update the waker cell
                    self.waker.with_mut(|t| *t = Some(new_waker));
```

Reply:
Cheers!